app/src/main/java/space/taran/arknavigator/mvp/presenter/adapter/ResourcesGridPresenter.kt

```diff
@@ -182,12 +206,72 @@ class ResourcesGridPresenter(
     ) = withContext(Dispatchers.Default) {
         this@ResourcesGridPresenter.resources = mapNewResources(resources)
         sortAllResources()
-        selection = this@ResourcesGridPresenter.resources
         withContext(Dispatchers.Main) {
             setProgressVisibility(false)
         }
     }
 
+    suspend fun shuffleResources() = withContext(Dispatchers.Default) {
+        val shuffledRes = selection
+            .filter {
+                !it.isPinned
+            }
+            .shuffled()
+        val unShuffledRes =
+            selection.filter { it.isPinned }
+                .toMutableList()
+        unShuffledRes.addAll(shuffledRes)
+        selection = unShuffledRes
+        selection.forEach {
+            Log.d("Resource id", it.meta.id.toString())
+        }
+        withContext(Dispatchers.Main) {
+            viewState.updateAdapter()
+        }
+    }
+
+    suspend fun increaseScore() = changeScore(1)
+
+    suspend fun decreaseScore() = changeScore(-1)
+
+    suspend fun resetScores() = withContext(Dispatchers.IO) {
+        scoreStorage.resetScores(selectedResources)
+        withContext(Dispatchers.Main) {
+            sortAllResources()
+            sortSelectionAndUpdateAdapter()
+            onSelectingChanged(false)
+        }
+    }
+
+    fun allowScoring() =
+        selectedResources.isNotEmpty() && selectedResources.size == 1 && sortByScores
```

Comment:
Why do we disallow scoring for multiple resources?

Reply:
I thought scoring should work for one resource every time. But that way, we can score many resources at once. I think that's ok